lib/src/Uncertainty/Bayesian/GenericMetropolisHastings.cxx

```diff
@@ -0,0 +1,157 @@
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "openturns/GenericMetropolisHastings.hxx"
+#include "openturns/PersistentObjectFactory.hxx"
+
+BEGIN_NAMESPACE_OPENTURNS
+
+
+CLASSNAMEINIT(GenericMetropolisHastings)
+
+static const Factory<GenericMetropolisHastings> Factory_GenericMetropolisHastings;
+
+/* Default constructor */
+GenericMetropolisHastings::GenericMetropolisHastings()
+  : MetropolisHastingsImplementation()
+{
+  // Nothing to do
+}
+
+
+/* Parameters constructor */
+GenericMetropolisHastings::GenericMetropolisHastings(const Distribution & targetDistribution,
+    const Point & initialState,
+    const Distribution & proposal,
+    const Function & linkFunction,
+    const Indices & marginalIndices)
+  : MetropolisHastingsImplementation(targetDistribution, initialState, marginalIndices)
+{
+  setProposalAndLinkFunction(proposal, linkFunction);
+}
+
+/* Parameters constructor */
+GenericMetropolisHastings::GenericMetropolisHastings(const Function & targetLogPDF,
+    const Domain & support,
+    const Point & initialState,
+    const Distribution & proposal,
+    const Function & linkFunction,
+    const Indices & marginalIndices)
+  : MetropolisHastingsImplementation(targetLogPDF, support, initialState, marginalIndices)
+{
+  setProposalAndLinkFunction(proposal, linkFunction);
+}
+
+
+/* String converter */
+String GenericMetropolisHastings::__repr__() const
+{
+  return OSS() << "class=" << GenericMetropolisHastings::GetClassName()
+         << " name=" << getName()
+         << " derived from " << MetropolisHastingsImplementation::__repr__()
+         << " proposal=" << proposal_;
+}
+
+
+GenericMetropolisHastings* GenericMetropolisHastings::clone() const
+{
+  return new GenericMetropolisHastings(*this);
+}
+
+
+Point GenericMetropolisHastings::getCandidate() const
+{
+  // proposal_ distribution conditioned on the currentState_
+  proposal_.setParameter(proposalLinkFunction_(currentState_));
+
+  const Point candidate(proposal_.getRealization());
+
+  // LogPDF of the proposal distribution at the candidate point
+  // conditional to the currentState_
+  const Scalar logProbNew = proposal_.computeLogPDF(candidate);
+
+  // Build potential new state from candidate
+  Point candidateState(currentState_);
+  for (UnsignedInteger j = 0; j < marginalIndices_.getSize(); ++ j)
+    candidateState[marginalIndices_[j]] = candidate[j];
+
+  // proposal_ distribution conditioned on the currentState_
```

Comment:
conditioned on the candidate state ?

Reply:
I dont really know what to say.

I dont like the name "Generic" :) maybe Proposal or Density/Distribution ? (we have RVMH)

Reply:
We called it "generic" because any MetropolisHastings class can be viewed as particular cases of this one. For instance:

- RandomWalkMetropolisHastings can be emulated with a GenericMetropolisHastings if the proposal distribution has an offset parameter equal to the current state of the chain (typically, the mean of a normal proposal distribution).
- RandomVectorMetropolisHastings can be emulated with a GenericMetropolisHastings by having the proposal distribution equal to the target distribution (this implies we are able to sample directly from the target distribution).
- IndependentMetropolisHastings can be emulated with a GenericMetropolisHastings by having a proposal distribution which does not depend on the state of the chain.

Other possible names would be "UserDefinedMetropolisHastings" or "GeneralMetropolisHastings". What do you think?